### PR TITLE
Fix: response.json() call makes unnecessary memory allocation

### PR DESCRIPTION
**Pull request by jxlil**
Now `response.json()` uses `response.body` to retrieve the `dict` object, so the str response is no longer stored in RAM.

Closes https://github.com/scrapy/scrapy/issues/5968

### CODE REVIEW

File: tests/test_http_response.py
Comment (jxlil) [inline tests/test_http_response.py:847]: Now,` json.loads()` receives bytes, so there's no need to use `decode()`

**Review by GeorgeA92 — APPROVED** [review]
From my side It looks OK.

However I am still not sure about.. additional tests that need (or not need) to implement here (I will prefer to wait for reviews from other contributors).

File: scrapy/http/response/text.py
Comment (kmike) [inline scrapy/http/response/text.py:85]: The behavior is not exactly the same, if the response is not UTF-8, and properly indicates it.

From https://datatracker.ietf.org/doc/html/rfc7159#section-8.1:

> JSON text SHALL be encoded in UTF-8, UTF-16, or UTF-32. 

From https://docs.python.org/3/library/json.html#json.loads:

> s can now be of type [bytes](https://docs.python.org/3/library/stdtypes.html#bytes) or [bytearray](https://docs.python.org/3/library/stdtypes.html#bytearray). The input encoding should be UTF-8, UTF-16 or UTF-32.

So we should be fine in most cases after this change. 

At the same time, SHALL in the RFC is not MUST (see https://datatracker.ietf.org/doc/html/rfc2119#section-3):

> SHOULD   This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course.

I'm fine with making the change, but we should add an entry to the changelog, which explains that the change is backwards incompatible, and that Scrapy stops support some of these weird cases to improve the efficiency for the vast majority of cases.

Comment (jxlil) [reply scrapy/http/response/text.py:85]: Done! I've just added an entry to the **Backward-incompatible changes** section in the changelog. Please let me know if I need to correct something.

File: docs/news.rst
Comment (Gallaecio) [inline docs/news.rst:54]: You are editing the release notes of the latest, already-released version of Scrapy.

We usually don’t edit release notes on PRs, and instead do it all at once during the release process. But this is actually a good idea, to make sure we do not forget about it.

So please remove this and add the following above the 2.10.0 header (I took the liberty to make some changes):

```rst
.. _release-2.11.0:

Scrapy 2.11.0 (to be released)
------------------------------

Backward-incompatible changes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

-   The :meth:`TextResponse.json <scrapy.http.TextResponse.json>` method now
    requires the response to be in a valid JSON encoding (UTF-8, UTF-16, or
    UTF-32).

    If you need to deal with JSON documents in an invalid encoding, use
    ``json.loads(response.text)`` instead.

    (:issue:`5968`)
```

**Review by Gallaecio — APPROVED** [review]
Looks good to me, provided the release notes fix.

File: docs/news.rst
Comment (jxlil) [reply docs/news.rst]: Thanks! The release notes are already updated

**Comment by jxlil** [comment]
A test failed, but I think it's not related to the changes in this PR

**Comment by kmike** [comment]
Thanks @jxlil!

**Pull request merged**
